### PR TITLE
[SPARK-25600][SQL][MINOR] Make use of TypeCoercion.findTightestCommonType while inferring CSV schema.

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Current the CSV's infer schema code inlines `TypeCoercion.findTightestCommonType`. This is a minor refactor to make use of the common type coercion code when applicable.  This way we can take advantage of any improvement to the base method.

Thanks to @MaxGekk for finding this while reviewing another PR.

## How was this patch tested?
This is a minor refactor.  Existing tests are used to verify the change.

### CODE REVIEW

**Comment by dilipbiswal** [comment]
cc @HyukjinKwon @MaxGekk

**Comment by ueshin** [comment]
Maybe this is related to #22448.

**Comment by dilipbiswal** [comment]
@gatorsmile There should not be any behaviour change. I was thinking that existing test cases should suffice. Basically we used to duplicate the code of TypeCoercion.findTightestCommonType in here. Here i am just reusing the common function.

**Comment by dilipbiswal** [comment]
@ueshin 
> Maybe this is related to #22448.
Yeah.. Actually @MaxGekk had pointed me to the presence of duplicate code in one of his comment. I was trying to address it in here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 1]
```diff
@@ -172,51 +172,35 @@ private[csv] object CSVInferSchema {
...
+        case (t1: DecimalType, t2: DecimalType) =>
+          val scale = math.max(t1.scale, t2.scale)
+          val range = math.max(t1.precision - t1.scale, t2.precision - t2.scale)
+          if (range + scale > 38) {
+            // DecimalType can't support precision > 38
+            Some(DoubleType)
+          } else {
+            Some(DecimalType(range + scale, scale))
+          }
+        case (_, _) => None
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: @HyukjinKwon Thanks. Will change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 2]
```diff
@@ -172,51 +172,35 @@ private[csv] object CSVInferSchema {
...
-      val scale = math.max(t1.scale, t2.scale)
-      val range = math.max(t1.precision - t1.scale, t2.precision - t2.scale)
-      if (range + scale > 38) {
-        // DecimalType can't support precision > 38
-        Some(DoubleType)
-      } else {
-        Some(DecimalType(range + scale, scale))
+  def compatibleType(t1: DataType, t2: DataType): Option[DataType] = {
+    TypeCoercion.findTightestCommonType(t1, t2).orElse {
+      (t1, t2) match {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: Can we leave this out as a private val like the previous and leave a comment that this pattern matching is CSV specific? That will reduce the diff and makes the review easier.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: BTW, let's keep the comments in the original place.

**Comment by HyukjinKwon** [comment]
Looks okay - I checked a case one by one but it needs another look.

**Comment by HyukjinKwon** [comment]
Let's just file a JIRA @dilipbiswal BTW.

**Comment by dilipbiswal** [comment]
@HyukjinKwon Okay.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 3]
```diff
@@ -172,51 +172,35 @@ private[csv] object CSVInferSchema {
...
-      Some(t1)
-
-    // These two cases below deal with when `IntegralType` is larger than `DecimalType`.
-    case (t1: IntegralType, t2: DecimalType) =>
-      findTightestCommonType(DecimalType.forType(t1), t2)
-    case (t1: DecimalType, t2: IntegralType) =>
-      findTightestCommonType(t1, DecimalType.forType(t2))
-
-    // Double support larger range than fixed decimal, DecimalType.Maximum should be enough
-    // in most case, also have better precision.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: Some comments here are ignored in the change. Shall we keep them?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 4]
```diff
@@ -172,51 +172,35 @@ private[csv] object CSVInferSchema {
...
-
-    case (t1: DecimalType, t2: DecimalType) =>
-      val scale = math.max(t1.scale, t2.scale)
-      val range = math.max(t1.precision - t1.scale, t2.precision - t2.scale)
-      if (range + scale > 38) {
-        // DecimalType can't support precision > 38
-        Some(DoubleType)
-      } else {
-        Some(DecimalType(range + scale, scale))
+  def compatibleType(t1: DataType, t2: DataType): Option[DataType] = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: nit: `findCompatibleType`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 3 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: @viirya Yeah.. we should keep.. sorry.. got dropped inadvertently.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 4 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: @viirya i kept the same name used in JsonInferSchema. Change that as well ? Or only change this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: @HyukjinKwon Did you have any preference or suggestion on the name of the val ? findCommonTypeExtended ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 4 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: `compatibleType` is also fine if it is consistent with `JsonInferSchema`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: not sure. maybe just `findCompatibleTypeForCSV`

**Review by HyukjinKwon — APPROVED** [review]

**Comment by dilipbiswal** [comment]
@HyukjinKwon Does this look okay now ?

**Comment by HyukjinKwon** [comment]
Yup. Let me leave this open few more days in case.

**Comment by dilipbiswal** [comment]
@HyukjinKwon Sure :-)

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**